File: VB/DataEntryFormSample/SpreadsheetBindingManager.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Linq
Imports System.Windows.Forms
Imports DevExpress.Spreadsheet
Imports DevExpress.XtraSpreadsheet

Namespace DataEntryFormSample
    ' This class is used to bind the data source object's properties to spreadsheet cells. 
    Partial Public Class SpreadsheetBindingManager
        Inherits Component

        Private _control As SpreadsheetControl
        Private _dataSource As Object
        Private currentItem As Object
        Private bindingManager As BindingManagerBase
        Private ReadOnly cellBindings As New Dictionary(Of String, String)()
        Private ReadOnly propertyDescriptors As New PropertyDescriptorCollection(Nothing)

        Public Sub New()
            InitializeComponent()
        End Sub

        Public Sub New(ByVal container As IContainer)
            container.Add(Me)
            InitializeComponent()
        End Sub

        Public Property Control() As SpreadsheetControl
            Get
                Return _control
            End Get
            Set(ByVal value As SpreadsheetControl)
                If Not ReferenceEquals(_control, value) Then
                    If _control IsNot Nothing Then
                        RemoveHandler _control.CellValueChanged, AddressOf SpreadsheetControl_CellValueChanged
                    End If
                    _control = value
                    If _control IsNot Nothing Then
                        AddHandler _control.CellValueChanged, AddressOf SpreadsheetControl_CellValueChanged
                    End If
                End If
            End Set
        End Property

        <Browsable(False), DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)>
        Public Property SheetName() As String

        Public Property DataSource() As Object
            Get
                Return _dataSource
            End Get
            Set(ByVal value As Object)
                If Not ReferenceEquals(_dataSource, value) Then
                    Detach()
                    _dataSource = value
                    Attach()
                End If
            End Set
        End Property

        ' Creates a binding between the data source object's property and a cell.
        ' propertyName specifies the data source property name.
        ' cellReference specifies a cell reference in the A1 reference style.
        Public Sub AddBinding(ByVal propertyName As String, ByVal cellReference As String)
            If cellBindings.ContainsKey(propertyName) Then
                Throw New ArgumentException($"Already has binding to {propertyName} property")
            End If
            Dim typedList As ITypedList = TryCast(_dataSource, ITypedList)
            If typedList IsNot Nothing Then
                Dim dataSourceProperties As PropertyDescriptorCollection = typedList.GetItemProperties(Nothing)
                Dim propertyDescriptor As PropertyDescriptor = dataSourceProperties(propertyName)
                If propertyDescriptor Is Nothing Then
                    Throw New InvalidOperationException($"Unknown { propertyName } property")
                End If
                If currentItem IsNot Nothing Then
                    propertyDescriptor.AddValueChanged(currentItem, AddressOf OnPropertyChanged)
                End If
                propertyDescriptors.Add(propertyDescriptor)
            End If
            cellBindings.Add(propertyName, cellReference)
        End Sub

        'Removes a binding for a data source property.
        'propertyName specifies the data source property name.</param>
        Public Sub RemoveBinding(ByVal propertyName As String)
            If cellBindings.ContainsKey(propertyName) Then
                Dim propertyDescriptor As PropertyDescriptor = propertyDescriptors(propertyName)
                If currentItem IsNot Nothing Then
                    propertyDescriptor.RemoveValueChanged(currentItem, AddressOf OnPropertyChanged)
                End If
                propertyDescriptors.Remove(propertyDescriptor)
                cellBindings.Remove(propertyName)
            End If
        End Sub

        'Removes all bindings.
        Public Sub ClearBindings()
            UnsubscribePropertyChanged()
            propertyDescriptors.Clear()
            cellBindings.Clear()
        End Sub

        ' Retrieves the binding manager and property descriptors, and subscribes to the data source and data members events.
        Private Sub Attach()
            Dim provider As ICurrencyManagerProvider = TryCast(_dataSource, ICurrencyManagerProvider)
            If provider IsNot Nothing Then
                bindingManager = provider.CurrencyManager
                AddHandler bindingManager.CurrentChanged, AddressOf BindingManager_CurrentChanged
                currentItem = bindingManager.Current
            End If
            Dim typedList As ITypedList = TryCast(_dataSource, ITypedList)
            If typedList IsNot Nothing Then
                Dim dataSourceProperties As PropertyDescriptorCollection = typedList.GetItemProperties(Nothing)
                For Each propertyName As String In cellBindings.Keys
                    Dim propertyDescriptor As PropertyDescriptor = dataSourceProperties(propertyName)
                    If propertyDescriptor Is Nothing Then
                        Throw New InvalidOperationException($"Unable to get property descriptor for { propertyName } property")
                    End If
                    propertyDescriptors.Add(propertyDescriptor)
                Next propertyName
            End If
            PullData()
            SubscribePropertyChanged()
        End Sub

        'Unsubscribes from the data source and data members events, and clears property descriptors. 
        Private Sub Detach()
            If _dataSource IsNot Nothing Then
                UnsubscribePropertyChanged()
                If bindingManager IsNot Nothing Then
                    RemoveHandler bindingManager.CurrentChanged, AddressOf BindingManager_CurrentChanged
                    bindingManager = Nothing
                    currentItem = Nothing
                End If
                propertyDescriptors.Clear()
            End If
        End Sub

        Private Sub BindingManager_CurrentChanged(ByVal sender As Object, ByVal e As EventArgs)
            _control?.BeginUpdate()
            Try
                'Update the data entry form when the current item changes.
                DeactivateCellEditor(CellEditorEnterValueMode.ActiveCell)
                UnsubscribePropertyChanged()
                currentItem = bindingManager.Current
                PullData()
                SubscribePropertyChanged()
                ActivateCellEditor()
            Finally
                _control?.EndUpdate()
            End Try
        End Sub

        Private Sub UnsubscribePropertyChanged()
            If currentItem IsNot Nothing Then
                For Each propertyDescriptor As PropertyDescriptor In propertyDescriptors
                    propertyDescriptor.RemoveValueChanged(currentItem, AddressOf OnPropertyChanged)
                Next propertyDescriptor
            End If
        End Sub

        Private Sub SubscribePropertyChanged()
            If currentItem IsNot Nothing Then
                For Each propertyDescriptor As PropertyDescriptor In propertyDescriptors
                    propertyDescriptor.AddValueChanged(currentItem, AddressOf OnPropertyChanged)
                Next propertyDescriptor
            End If
        End Sub

        Private Sub OnPropertyChanged(ByVal sender As Object, ByVal eventArgs As EventArgs)

            'Update the bound cell's value when the corresponding data source property's value changes.
            Dim propertyDescriptor As PropertyDescriptor = TryCast(sender, PropertyDescriptor)
            If propertyDescriptor IsNot Nothing AndAlso currentItem IsNot Nothing Then
                Dim reference As String = Nothing
                If cellBindings.TryGetValue(propertyDescriptor.Name, reference) Then
                    SetCellValue(reference, CellValue.FromObject(propertyDescriptor.GetValue(currentItem)))
                End If
            End If
        End Sub

        'Pull data from the data source (update values of all bound cells).
        Private Sub PullData()
            If currentItem IsNot Nothing Then
                For Each propertyDescriptor As PropertyDescriptor In propertyDescriptors
                    Dim reference As String = cellBindings(propertyDescriptor.Name)
                    SetCellValue(reference, CellValue.FromObject(propertyDescriptor.GetValue(currentItem)))
                Next propertyDescriptor
            End If
        End Sub

        Private Sub SpreadsheetControl_CellValueChanged(ByVal sender As Object, ByVal e As SpreadsheetCellEventArgs)

            'Update the data source property's value when the bound cell's value changes.
            If e.SheetName = SheetName Then
                Dim reference As String = e.Cell.GetReferenceA1()
                Dim propertyName As String = cellBindings.SingleOrDefault(Function(p) p.Value = reference).Key
                If Not String.IsNullOrEmpty(propertyName) Then
                    Dim propertyDescriptor As PropertyDescriptor = propertyDescriptors(propertyName)
                    If propertyDescriptor IsNot Nothing AndAlso currentItem IsNot Nothing Then
                        propertyDescriptor.SetValue(currentItem, e.Value.ToObject())
                    End If
                End If
            End If
        End Sub

        Private ReadOnly Property Sheet() As Worksheet
            Get
                Return If(_control IsNot Nothing AndAlso _control.Document.Worksheets.Contains(SheetName), _control.Document.Worksheets(SheetName), Nothing)
            End Get
        End Property

        Private Sub SetCellValue(ByVal reference As String, ByVal value As CellValue)
            If Sheet IsNot Nothing Then
                If reference = Sheet.Selection.GetReferenceA1() Then
                    DeactivateCellEditor()
                End If
                Sheet(reference).Value = value
                If reference = Sheet.Selection.GetReferenceA1() Then
                    ActivateCellEditor()
                End If
            End If
        End Sub

        Private Sub ActivateCellEditor()
            
            Dim _sheet = Sheet
            If _sheet IsNot Nothing Then
                Dim editors = _sheet.CustomCellInplaceEditors.GetCustomCellInplaceEditors(_sheet.Selection)
                If editors.Count = 1 Then
                    _control.OpenCellEditor(CellEditorMode.Edit)
                End If
            End If
        End Sub

        Private Sub DeactivateCellEditor(Optional ByVal mode As CellEditorEnterValueMode = CellEditorEnterValueMode.Cancel)
            If _control IsNot Nothing AndAlso _control.IsCellEditorActive Then
                _control.CloseCellEditor(mode)
            End If
        End Sub
    End Class
End Namespace

File: CS/DataEntryFormSample/SpreadsheetBindingManager.cs
using DevExpress.Spreadsheet;
using DevExpress.XtraSpreadsheet;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace DataEntryFormSample {
    /// <summary>
    /// This class is used to bind the data source object's properties to spreadsheet cells. 
    /// </summary>
    public partial class SpreadsheetBindingManager : Component {
        private SpreadsheetControl control;
        private object dataSource;
        private object currentItem;
        private BindingManagerBase bindingManager;
        private readonly Dictionary<string, string> cellBindings = new Dictionary<string, string>();
        private readonly PropertyDescriptorCollection propertyDescriptors = new PropertyDescriptorCollection(null);

        public SpreadsheetBindingManager() {
            InitializeComponent();
        }

        public SpreadsheetBindingManager(IContainer container) {
            container.Add(this);
            InitializeComponent();
        }

        public SpreadsheetControl Control {
            get => control;
            set {
                if (!ReferenceEquals(control, value)) {
                    if (control != null)
                        control.CellValueChanged -= SpreadsheetControl_CellValueChanged;
                    control = value;
                    if (control != null)
                        control.CellValueChanged += SpreadsheetControl_CellValueChanged;
                }
            }
        }

        [Browsable(false)]
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
        public string SheetName { get; set; }

        public object DataSource {
            get => dataSource;
            set {
                if (!ReferenceEquals(dataSource, value)) {
                    Detach();
                    dataSource = value;
                    Attach();
                }
            }
        }

        /// <summary>
        /// Creates a binding between the data source object's property and a cell.
        /// </summary>
        /// <param name="propertyName">Specifies the data source property name.</param>
        /// <param name="cellReference">Specifies a cell reference in the A1 reference style.</param>
        public void AddBinding(string propertyName, string cellReference) {
            if (cellBindings.ContainsKey(propertyName))
                throw new ArgumentException($"A binding to the {propertyName} property already exists");
            if (dataSource is ITypedList typedList) {
                PropertyDescriptorCollection dataSourceProperties = typedList.GetItemProperties(null);
                PropertyDescriptor propertyDescriptor = dataSourceProperties[propertyName];
                if (propertyDescriptor == null)
                    throw new InvalidOperationException($"Unknown { propertyName } property");
                if (currentItem != null)
                    propertyDescriptor.AddValueChanged(currentItem, OnPropertyChanged);
                propertyDescriptors.Add(propertyDescriptor);
            }
            cellBindings.Add(propertyName, cellReference);
        }

        /// <summary>
        /// Removes a binding for a data source property.
        /// </summary>
        /// <param name="propertyName">Specifies the data source property name.</param>
        public void RemoveBinding(string propertyName) {
            if (cellBindings.ContainsKey(propertyName)) {
                PropertyDescriptor propertyDescriptor = propertyDescriptors[propertyName];
                if (currentItem != null)
                    propertyDescriptor.RemoveValueChanged(currentItem, OnPropertyChanged);
                propertyDescriptors.Remove(propertyDescriptor);
                cellBindings.Remove(propertyName);
            }
        }

        /// <summary>
        /// Removes all bindings.
        /// </summary>
        public void ClearBindings() {
            UnsubscribePropertyChanged();
            propertyDescriptors.Clear();
            cellBindings.Clear();
        }

        /// <summary>
        /// Retrieves the binding manager and property descriptors, and subscribes to the data source and data members events.
        /// </summary>
        private void Attach() {
            if (dataSource is ICurrencyManagerProvider provider) {
                bindingManager = provider.CurrencyManager;
                bindingManager.CurrentChanged += BindingManager_CurrentChanged;
                currentItem = bindingManager.Current;
            }
            if (dataSource is ITypedList typedList) {
                PropertyDescriptorCollection dataSourceProperties = typedList.GetItemProperties(null);
                foreach (string propertyName in cellBindings.Keys) {
                    PropertyDescriptor propertyDescriptor = dataSourceProperties[propertyName];
                    if (propertyDescriptor == null)
                        throw new InvalidOperationException($"Unable to get a property descriptor for the { propertyName } property");
                    propertyDescriptors.Add(propertyDescriptor);
                }
            }
            PullData();
            SubscribePropertyChanged();
        }

        /// <summary>
        /// Unsubscribes from the data source and data members events, and clears property descriptors. 
        /// </summary>
        private void Detach() {
            if (dataSource != null) {
                UnsubscribePropertyChanged();
                if (bindingManager != null) {
                    bindingManager.CurrentChanged -= BindingManager_CurrentChanged;
                    bindingManager = null;
                    currentItem = null;
                }
                propertyDescriptors.Clear();
            }
        }

        private void BindingManager_CurrentChanged(object sender, EventArgs e) {
            // Update the data entry form when the current item changes.
            DeactivateCellEditor(CellEditorEnterValueMode.ActiveCell);
            control?.BeginUpdate();
            try {
                UnsubscribePropertyChanged();
                currentItem = bindingManager.Current;
                PullData();
                SubscribePropertyChanged();
            }
            finally {
                control?.EndUpdate();
                ActivateCellEditor();
            }
        }

        private void UnsubscribePropertyChanged() {
            if (currentItem != null) {
                foreach (PropertyDescriptor propertyDescriptor in propertyDescriptors)
                    propertyDescriptor.RemoveValueChanged(currentItem, OnPropertyChanged);
            }
        }
        private void SubscribePropertyChanged() {
            if (currentItem != null) {
                foreach (PropertyDescriptor propertyDescriptor in propertyDescriptors)
                    propertyDescriptor.AddValueChanged(currentItem, OnPropertyChanged);
            }
        }

        private void OnPropertyChanged(object sender, EventArgs eventArgs) {
            // Update the bound cell's value when the corresponding data source property's value changes.
            PropertyDescriptor propertyDescriptor = sender as PropertyDescriptor;
            if (propertyDescriptor != null && currentItem != null) {
                string reference;
                if (cellBindings.TryGetValue(propertyDescriptor.Name, out reference))
                    SetCellValue(reference, CellValue.FromObject(propertyDescriptor.GetValue(currentItem)));
            }
        }

        // Pull data from the data source (update values of all bound cells).
        private void PullData() {
            if (currentItem != null) {
                foreach (PropertyDescriptor propertyDescriptor in propertyDescriptors) {
                    string reference = cellBindings[propertyDescriptor.Name];
                    SetCellValue(reference, CellValue.FromObject(propertyDescriptor.GetValue(currentItem)));
                }
            }
        }

        private void SpreadsheetControl_CellValueChanged(object sender, SpreadsheetCellEventArgs e) {
            // Update the data source property's value when the bound cell's value changes.
            if (e.SheetName == SheetName) {
                string reference = e.Cell.GetReferenceA1();
                string propertyName = cellBindings.SingleOrDefault(p => p.Value == reference).Key;
                if (!string.IsNullOrEmpty(propertyName)) {
                    PropertyDescriptor propertyDescriptor = propertyDescriptors[propertyName];
                    if (propertyDescriptor != null && currentItem != null)
                        propertyDescriptor.SetValue(currentItem, e.Value.ToObject());
                }
            }
        }

        private Worksheet Sheet => 
            control != null && control.Document.Worksheets.Contains(SheetName) ? control.Document.Worksheets[SheetName] : null;

        private void SetCellValue(string reference, CellValue value) {
            if (Sheet != null) {
                if (reference == Sheet.Selection.GetReferenceA1())
                    DeactivateCellEditor();
                Sheet[reference].Value = value;
                if (reference == Sheet.Selection.GetReferenceA1())
                    ActivateCellEditor();
            }
        }

        private void ActivateCellEditor() {
            var sheet = Sheet;
            if (sheet != null) {
                var editors = sheet.CustomCellInplaceEditors.GetCustomCellInplaceEditors(sheet.Selection);
                if (editors.Count == 1)
                    control.OpenCellEditor(CellEditorMode.Edit);
            }
        }

        private void DeactivateCellEditor(CellEditorEnterValueMode mode = CellEditorEnterValueMode.Cancel) {
            if (control != null && control.IsCellEditorActive)
                control.CloseCellEditor(mode);
        }
    }
}
